text_extensions_for_pandas/transformers/tokenization.py

```diff
@@ -0,0 +1,90 @@
+#
+#  Copyright (c) 2020 IBM Corp.
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+
+################################################################################
+# tokenize.py
+#
+# Functions for tokenization of text
+
+import numpy as np
+import pandas as pd
+from transformers.tokenization_utils import PreTrainedTokenizerFast
+
+from text_extensions_for_pandas.array import (
+    CharSpanArray,
+    TokenSpanArray,
+)
+
+
+def make_bert_tokens(target_text: str,
+                     tokenizer: PreTrainedTokenizerFast
+                     ) -> pd.DataFrame:
+    """
+    Tokenize the indicated text for BERT embeddings and return a DataFrame
+    with one row per token.
+
+    :param: target_text: string to tokenize
+    :param: tokenizer: A tokenizer that is a subclass of huggingface transformers
+                       PreTrainingTokenizerFast that supports `encode_plus`.
+
+    :returns: A `pd.DataFrame` with the following columns:
+     * "id": unique integer ID for each token
+     * "char_span": span of the token with character offsets
+     * "token_span": span of the token with token offsets
+     * "input_id": integer ID suitable for input to a BERT embedding model
+     * "token_type_id": list of token type ids to be fed to a model
+     * "attention_mask": list of indices specifying which tokens should be
+                         attended to by the model
+     * "special_tokens_mask": `True` if the token is a zero-length special token
+       such as "start of document"
+    """
+    tokenized_result = tokenizer.encode_plus(target_text,
+                                             return_special_tokens_mask=True,
+                                             return_offsets_mapping=True)
+    df = pd.DataFrame.from_dict(tokenized_result)
+    # Get offset mapping from tokenizer
+    offsets = tokenized_result["offset_mapping"]
+
+    # Turn special tokens into zero-length spans
+    begins = np.zeros(len(df.index), dtype=np.int64)
+    ends = np.zeros(len(df.index), dtype=np.int64)
+    prev_begin = 0
+    for i in range(len(df.index)):
```

Comment:
Can you vectorize this loop?
You should be able to implement this transformation as follows: 
1. Correct the `begin` values for any special tokens at the very beginning of the text (either set them to 0, or set them to the begin offset of the first non-special token).
2. Use [`pandas.DataFrame.fillna`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html#pandas.DataFrame.fillna) with the `'ffill'` imputation method to fill in the missing `begin`.
3. Replace missing `end` values with the corresponding value from the corrected `begin` series.